### PR TITLE
Fix: resolve nested attributes of anonymous records in GetDatamembers

### PR DESCRIPTION
**Pull request by Vipul-Cariappa**

### CODE REVIEW

**Comment by Vipul-Cariappa** [comment]
Fixes: https://github.com/Vipul-Cariappa/cppyy-compiler-research/blob/9a73ba6a0cda758991827e51141f3893002dcae5/test/test_datatypes.py#L1562-L1563 and https://github.com/Vipul-Cariappa/cppyy-compiler-research/blob/9a73ba6a0cda758991827e51141f3893002dcae5/test/test_datatypes.py#L1656-L1657 without any extra modification.

File: lib/Interpreter/CppInterOp.cpp [hunk 1]
```diff
@@ -1128,9 +1128,33 @@
...
+      stack_end.push_back(CXXRD->field_end());
+      while (stack_begin.size()) {
+        if (stack_begin.back() == stack_end.back()) {
+          stack_begin.pop_back();
+          stack_end.pop_back();
+          continue;
+        }
+        Decl* D = *(stack_begin.back());
+        if (D->getKind() == clang::Decl::Field &&
+            dyn_cast<FieldDecl>(D)->isAnonymousStructOrUnion()) {
```
[the remark this answers is not in the document]
Comment (Vipul-Cariappa) [reply lib/Interpreter/CppInterOp.cpp:1143]: It cannot be null as we are iterating through `FieldDecl`s. The `dyn_cast` cannot return null.

File: lib/Interpreter/CppInterOp.cpp [hunk 2]
```diff
@@ -1128,9 +1128,33 @@
...
+          if (auto* CXXRD = llvm::dyn_cast_or_null<CXXRecordDecl>(
+                  dyn_cast<FieldDecl>(D)
```
[the remark this answers is not in the document]
Comment (Vipul-Cariappa) [reply lib/Interpreter/CppInterOp.cpp:1145]: same as above

File: lib/Interpreter/CppInterOp.cpp [hunk 3]
```diff
@@ -1128,9 +1128,33 @@ namespace Cpp {
...
+      stack_end.push_back(CXXRD->field_end());
+      while (!stack_begin.empty()) {
+        if (stack_begin.back() == stack_end.back()) {
+          stack_begin.pop_back();
+          stack_end.pop_back();
+          continue;
+        }
+        Decl* D = *(stack_begin.back());
+        if (D->getKind() == clang::Decl::Field &&
+            dyn_cast<FieldDecl>(D)->isAnonymousStructOrUnion()) {
```
Comment (vgvassilev) [inline lib/Interpreter/CppInterOp.cpp:1143]:
```suggestion
        if (isa<FieldDecl>(D) &&
            cast<FieldDecl>(D)->isAnonymousStructOrUnion()) {
```

File: lib/Interpreter/CppInterOp.cpp [hunk 4]
```diff
@@ -1175,9 +1199,30 @@
...
     auto &C = getASTContext();
 
-    if (auto *FD = llvm::dyn_cast<FieldDecl>(D))
-      return (intptr_t) C.toCharUnitsFromBits(C.getASTRecordLayout(FD->getParent())
-                                              .getFieldOffset(FD->getFieldIndex())).getQuantity();
+    if (auto* FD = llvm::dyn_cast<FieldDecl>(D)) {
+      const clang::RecordDecl* RD = FD->getParent();
+      intptr_t offset =
+          C.toCharUnitsFromBits(
+               C.getASTRecordLayout(RD).getFieldOffset(FD->getFieldIndex()))
```
[the remark this answers is not in the document]
Comment (Vipul-Cariappa) [reply lib/Interpreter/CppInterOp.cpp:1206]: We were doing the same thing before. It should be safe according to me.

File: lib/Interpreter/CppInterOp.cpp [hunk 5]
```diff
@@ -1175,9 +1199,30 @@
...
+        for (auto f = RD->field_begin(); f != RD->field_end(); ++f) {
+          auto* rt = f->getType()->getAs<RecordType>();
+          if (!rt)
+            continue;
+          if (anon == rt->getDecl()) {
+            FD = *f;
+            break;
+          }
+        }
+        offset += C.toCharUnitsFromBits(C.getASTRecordLayout(RD).getFieldOffset(
```
[the remark this answers is not in the document]
Comment (Vipul-Cariappa) [reply lib/Interpreter/CppInterOp.cpp:1220]: same as above

File: lib/Interpreter/CppInterOp.cpp (hunk 3 above)
Comment (vgvassilev) [reply lib/Interpreter/CppInterOp.cpp:1143]: I think we should have a proper dyn_cast and `isAnonymousStructOrUnion` as a nested if...

**Comment by vgvassilev** [comment]
> The [test fail](https://github.com/compiler-research/CppInterOp/actions/runs/10825793390/job/30035482367?pr=321#step:15:204) with `ISO C++ prohibits anonymous structs [-Werror=pedantic]`, and one of the tests has an anonymous struct. We have a similar anonymous struct in the [cppyy tests](https://github.com/Vipul-Cariappa/cppyy-compiler-research/blob/9a73ba6a0cda758991827e51141f3893002dcae5/test/test_datatypes.py#L1611-L1615). Should I add `-Wno-pedantic` to the compiler flags?

Yes, but only for this test file.

**Comment by Vipul-Cariappa** [comment]
Hi @vgvassilev, [Valgrind fails](https://github.com/compiler-research/CppInterOp/actions/runs/10832284707/job/30056328026?pr=321#step:15:883). Should I add
```c++
if (llvm::sys::RunningOnValgrind())
    GTEST_SKIP() << "XFAIL due to Valgrind report";
```
?

It is `Conditional jump or move depends on uninitialised value(s)` at LLVM's code base, according to my eyes.

File: lib/Interpreter/CppInterOp.cpp [hunk 6]
```diff
@@ -1128,9 +1128,32 @@ namespace Cpp {
...
+      stack_begin.push_back(CXXRD->field_begin());
+      stack_end.push_back(CXXRD->field_end());
+      while (!stack_begin.empty()) {
+        if (stack_begin.back() == stack_end.back()) {
+          stack_begin.pop_back();
+          stack_end.pop_back();
+          continue;
+        }
+        Decl* D = *(stack_begin.back());
+        if (D->getKind() == clang::Decl::Field) {
```
Comment (vgvassilev) [inline lib/Interpreter/CppInterOp.cpp:1142]:
```suggestion
```

File: lib/Interpreter/CppInterOp.cpp [hunk 7]
```diff
@@ -1175,9 +1198,30 @@ namespace Cpp {
...
     auto &C = getASTContext();
 
-    if (auto *FD = llvm::dyn_cast<FieldDecl>(D))
-      return (intptr_t) C.toCharUnitsFromBits(C.getASTRecordLayout(FD->getParent())
-                                              .getFieldOffset(FD->getFieldIndex())).getQuantity();
+    if (auto* FD = llvm::dyn_cast<FieldDecl>(D)) {
+      const clang::RecordDecl* RD = FD->getParent();
+      intptr_t offset =
+          C.toCharUnitsFromBits(
+               C.getASTRecordLayout(RD).getFieldOffset(FD->getFieldIndex()))
```
Comment (vgvassilev) [inline lib/Interpreter/CppInterOp.cpp:1205]:
```suggestion
               C.getFieldOffset(FD)
```

**Comment by vgvassilev** [comment]
> Hi @vgvassilev, [Valgrind fails](https://github.com/compiler-research/CppInterOp/actions/runs/10832284707/job/30056328026?pr=321#step:15:883). Should I add
> 
> ```c++
> if (llvm::sys::RunningOnValgrind())
>     GTEST_SKIP() << "XFAIL due to Valgrind report";
> ```
> 
> ?
> 
> It is `Conditional jump or move depends on uninitialised value(s)` at LLVM's code base, according to my eyes.

Yes, please do.

File: lib/Interpreter/CppInterOp.cpp [hunk 8]
```diff
@@ -1128,9 +1128,32 @@ namespace Cpp {
...
+      stack_begin.push_back(CXXRD->field_begin());
+      stack_end.push_back(CXXRD->field_end());
+      while (!stack_begin.empty()) {
+        if (stack_begin.back() == stack_end.back()) {
+          stack_begin.pop_back();
+          stack_end.pop_back();
+          continue;
+        }
+        Decl* D = *(stack_begin.back());
+        if (auto* FD = llvm::dyn_cast_or_null<FieldDecl>(D)) {
```
Comment (vgvassilev) [inline lib/Interpreter/CppInterOp.cpp:1142]:
```suggestion
        if (auto* FD = llvm::dyn_cast<FieldDecl>(D)) {
```

File: lib/Interpreter/CppInterOp.cpp [hunk 9]
```diff
@@ -1128,9 +1128,32 @@ namespace Cpp {
...
+          if (FD->isAnonymousStructOrUnion()) {
+            if (auto* RT = FD->getType()->getAs<RecordType>()) {
+              if (auto* CXXRD =
+                      llvm::dyn_cast_or_null<CXXRecordDecl>(RT->getDecl())) {
```
Comment (vgvassilev) [inline lib/Interpreter/CppInterOp.cpp:1146]:
```suggestion
              if (auto* CXXRD =
                      llvm::dyn_cast<CXXRecordDecl>(RT->getDecl())) {
```

**Review by vgvassilev — APPROVED** [review]
Lgtm!

**Pull request merged**